### ISSUE TITLE
Change ArrayBuffer arguments to ArrayBuffer + srcOffset

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#244]
WebGL added entry points which take an ArrayBuffer + srcOffset (offset into the AB). This was necessary so that applications don't have to constantly create ArrayBufferView garbage for every call. Especially important for WASM where the source AB is WASM memory - so always requires an offset.

Could be a dictionary that has AB+offset, I suppose, but likely creates just as much garbage.

No one likes that we have to design around GC, but it's how it is.

**Comment by Kangz** [comment]
Closing. All methods taking an `ArrayBuffer` (or `BufferSource`): `writeTexture` and `writeBuffer` also take an offset (in `GPUTextureDataLayout` for `writeTexture`).

**Comment by domenic** [comment]
I wish there were benchmarks supporting this performance claim about ArrayBufferViews creating too much garbage.

**Comment by kainino0x** [comment]
> I wish there were benchmarks supporting this performance claim about ArrayBufferViews creating too much garbage.

@juj do you happen to have any data for this?

**Comment by juj** [comment]
Two data points from the past that I can recall:
1. Benchmarking the adoption of [WebGL 2 garbage free entry points](https://github.com/emscripten-core/emscripten/pull/4743) improved performance in a real world rendering demo Sun Temple in Unreal Engine 4 by a high single digit number (~+8% iirc). There should be a thread with numbers either at emscripten github or emscripten-discuss mailing list, but was unable to find that now.
2. Benchmarking the optimization of [pooling glUniform*iv calls](https://github.com/emscripten-core/emscripten/pull/9953) improved performance in a real world application in bgfx engine by +17.6%.

Additionally I recall a third scenario related to per-frame work in Web Audio API where we optimized temporary GC sources, and found that to both improve performance and fix microstuttering that would happen every few seconds.

> I wish there were benchmarks supporting this performance claim about ArrayBufferViews creating too much garbage.

The wording suggests that you are skeptical that it would have any difference.

A major difficulty still with each browser is that they do not have good profilers to do more fine grained performance profiling than at the function level. Also people are generally disinterested in small % wins, thinking those are negligible. That is why these types of issues are either ignored or hidden from sight.

For the purposes of this conversation, the data points above are probably not conclusive to convince a skeptic, they were never written to back an expectation of having a "burden of proof".

I don't mind if WebGPU first does the webby JS thing and then in a future generation of the API revises the API to optimize for a lean wasm->js interop performance. That way there will be data. Currently the interop is quite verbose that gc pressure will not likely show up as the highest bottleneck - but I am yet to port a large wasm renderer to profile that hunch.

**Comment by domenic** [comment]
Yes, I think something that actually benchmarked the APIs in question, and in particular wasn't a general "garbage vs. no garbage" benchmark but instead an "ArrayBufferView vs. multiple parameters" benchmark, would be helpful, before deciding to depart from the idiomatic JS API.

**Comment by kainino0x** [comment]
Opening to collect data. Chromium (at least) implements both already, we should be able to easily write an artificial benchmark.

**Comment by domenic** [comment]
Great. A non-artificial benchmark would be even better, e.g. a demo or game that you could measure the FPS of.

**Comment by litherum** [comment]
+1 for benchmarking

**Comment by kainino0x** [comment]
FWIW the polyfill for these overloads is actually quite complicated due to the barebones-ness of the TypedArray APIs, especially because TypedArrays can't be made directly from other TypedArrays.

`f(data, dataOffset, size)`
becomes
```ts
if ('BYTES_PER_ELEMENT' in data) {
  f(new Uint8Array(data.buffer,
      data.byteOffset + (dataOffset ?? 0) * data.BYTES_PER_ELEMENT,
      (size ?? data.length) * data.BYTES_PER_ELEMENT));
} else {
  f(new Uint8Array(data, dataOffset, size));
}
```
(I'm not even totally sure this is correct.)

> Great. A non-artificial benchmark would be even better, e.g. a demo or game that you could measure the FPS of.

I somewhat disagree actually. A microbenchmark could easily tell us if there is no measurable difference at all, which would inform how to approach further benchmarking. (And I think this is a real possibility with today's JIT compilers.) And in fact, ideally the microbenchmark wouldn't use WebGL or WebGPU, as that could also easily confound the test.

I think a real-world benchmark would be sensitive to hundreds of details in exactly how the benchmark is built, each of which could easily hide a real performance issue (behind various bottlenecks or latencies), making the benchmark entirely misleading. For example: garbage cost hiding within other garbage cost; CPU cost hidden by GPU bottleneck; CPU cost of marking hidden by latencies of GPU operations, browser IPC; etc.

**Comment by domenic** [comment]
I would suggest making this a priority part of TAG review since it is such a departure from the idiomatic pattern we have recommended and standardized on for other web APIs.

**Comment by Kangz** [comment]
@juj posted these numbers about previous experience with this type of entrypoint:

> 1. Benchmarking the adoption of https://github.com/emscripten-core/emscripten/pull/4743 improved performance in a real world rendering demo Sun Temple in Unreal Engine 4 by a high single digit number (~+8% iirc). There should be a thread with numbers either at emscripten github or emscripten-discuss mailing list, but was unable to find that now.
> 2. Benchmarking the optimization of https://github.com/emscripten-core/emscripten/pull/9953 improved performance in a real world application in bgfx engine by +17.6%.

and the polyfill @kainino0x produced showed that there's necessarily garbage being created so we can't separate an "ArrayBufferView vs. multiple parameters" benchmark from "garbage vs. no garbage" one.

Aren't these two elements sufficient to motivate the overloads? We can go to the TAG but the reviews have been taking forever (I actually have a TODO list item to "celebrate" the 1-year birthday of that review request tomorrow).

**Comment by kainino0x** [comment]
editors likely to move this back to tacit resolution (to close without change) unless there's further feedback.

**Comment by juj** [comment]
What are the actual entry points that are being discussed here? When I look at my https://github.com/juj/wasm_webgpu library, I don't see this being a source of trash, since the API that currently exists and deals with buffers does already do the Fast Thing(tm).

There is

```webidl
interface mixin GPUBindingCommandsMixin {

    undefined setBindGroup(GPUIndex32 index, GPUBindGroup bindGroup,
        Uint32Array dynamicOffsetsData,
        GPUSize64 dynamicOffsetsDataStart,
        GPUSize32 dynamicOffsetsDataLength);
};
```

which already has offset+length access to the data array, so it is garbage-free to interop with wasm:
 - [GPUBindingCommandsMixin.setBindGroup](https://github.com/juj/wasm_webgpu/blob/ef0af80b998e6c703eadb036789d18a5972f4ae5/lib/lib_webgpu.js#L1764-L1771),

Then there's

```webidl
interface GPUQueue {
    undefined writeBuffer(
        GPUBuffer buffer,
        GPUSize64 bufferOffset,
        [AllowShared] BufferSource data,
        optional GPUSize64 dataOffset = 0,
        optional GPUSize64 size);
```

which also is garbage-free: [GPUQueue.writeBuffer](https://github.com/juj/wasm_webgpu/blob/35156a25b27d12156d982c3e65d4ea6f060bd430/lib/lib_webgpu.js#L1870-L1876)

Finally, I find a third function
```webidl
    undefined writeTexture(
        GPUImageCopyTexture destination,
        [AllowShared] BufferSource data,
        GPUImageDataLayout dataLayout,
        GPUExtent3D size);
```
that also has a source+offset kind of structure, so does not suffer from temp garbage from that source, although it does generate temp garbage due to the descriptor like nature of the function parameters: [GPUQueue.writeTexture](https://github.com/juj/wasm_webgpu/blob/35156a25b27d12156d982c3e65d4ea6f060bd430/lib/lib_webgpu.js#L1879-L1885)

Beyond the scope of this ticket, there is

```webidl
interface GPUBuffer {
    Promise<undefined> mapAsync(GPUMapModeFlags mode, optional GPUSize64 offset = 0, optional GPUSize64 size);
    ArrayBuffer getMappedRange(optional GPUSize64 offset = 0, optional GPUSize64 size);
```
which does trash since a new `ArrayBuffer` is returned.

I wonder if I missed some API entry points that aren't doing ArrayBuffer + srcOffset atm?

**Comment by juj** [comment]
The current sources of unavoidable **typed array**-related trash that I am observing are these:

https://github.com/juj/wasm_webgpu/blob/35156a25b27d12156d982c3e65d4ea6f060bd430/lib/lib_webgpu.js#L793-L795

https://github.com/juj/wasm_webgpu/blob/35156a25b27d12156d982c3e65d4ea6f060bd430/lib/lib_webgpu.js#L806-L808

That is, in the current ArrayBuffer specification, it is not possible to copy data between the Wasm memory heap and the `ArrayBuffer` returned by `GPUBuffer.getMappedRange()` due to the WhatWG HTML spec missing a suitable ArrayBuffer/TypedArray function that would enable garbage-free copying of data between two BufferSources.

------------

Taking a step back, it should be noted here that the overall temp garbage issue that WebGL had happened to be about instantiating temp ArrayBufferViews, so removing them was what was needed for WebGL to make it garbage free.

WebGPU on the other hand is temp trashy on **many** other places as well, not just around ArrayBufferViews. So optimizing (only) what optimized WebGL wouldn't fix WebGPU to be temp garbage free. The descriptor-based design generates garbage in WebGPU, and so does buffer mapping, render bundle encoder creation, query creation, starting render passes, creating bind groups and doing texture-buffer copies, at a glance through the spec.

So to fix WebGPU garbage issue, different optimizations would be needed than what WebGL needed.

There is some hope for WebGPU though: in WebGL the sources of temp garbage were quite tight close to the per-object draw loop, whereas in WebGPU the sources of garbage are in somewhat colder call sites. Unfortunately we don't have engines ported and optimized for WebGPU yet to be able to benchmark how much the slightly more cold paths of temp garbage will help mitigate this issue.

**Comment by kainino0x** [comment]
> I wonder if I missed some API entry points that aren't doing ArrayBuffer + srcOffset atm?

We already added the `srcOffset`s to all of the relevant entry points in WebGPU - that was the first time this issue got closed. It got reopened because we were asked to consider removing them again.

With regard to garbage in general (both temporary descriptors and ArrayBufferViews), I think there's some hope that today's JS engines are good enough at escape analysis that they can actually allocate these objects on the stack and clean them up immediately. This is yet to be tested though.
